INSP: add RsRedundantElseInspection

Kobzol opened the pull request:
Adds inspection to find redundant else branches, as discussed in https://github.com/intellij-rust/intellij-rust/pull/4972.

Right now it doesn't work for these irrefutable patterns:
```rust
enum E { V1 }
fn main() {
    let a = E::V1;
    if let E::V1 = a {
    } else { // <- this else should be redundant
    }
}
```
because `let E::V1 = a` is `RsPatConst` and `RsPat.isIrrefutable` returns `false` for it. Should I change this?

Should the inspection's warning range cover only `else` or the whole block after the `else`?

It would be also nice if it recognized irrefutable conditions like `if true` or `if 5 == 5`, but we need some basic constant propagation for this to work (is there something like this in the plugin already)?

mchernyavsky commented on src/main/kotlin/org/rust/ide/inspections/RsRedundantElseInspection.kt:
```diff
@@ -0,0 +1,59 @@
+                val elseExpr = expr.`else`
+                holder.registerProblem(
+                    expr,
+                    TextRange(elseExpr.startOffsetInParent, elseExpr.endOffsetInParent),
```
```suggestion
                    elseExpr.textRangeInParent,
```

mchernyavsky commented on src/main/kotlin/org/rust/ide/inspections/RsRedundantElseInspection.kt:
```diff
@@ -0,0 +1,59 @@
+            }
+        }
+
+    private fun RsElseBranch.isRedundant(): Boolean {
+        val set = mutableSetOf<RsCondition>()
+        var candidate: PsiElement = this
+
+        while (candidate is RsElseBranch || candidate is RsIfExpr) {
+            candidate.leftSiblings.filterIsInstance<RsCondition>().forEach { set.add(it) }
+            candidate = candidate.parent
+        }
+
+        return set.any {
+            it.orPats?.patList?.all { pat -> pat?.isIrrefutable ?: false } ?: false
+        }
+    }
```
```suggestion
    companion object {
        private val RsElseBranch.isRedundant: Boolean
            get() {
                val set = mutableSetOf<RsCondition>()
                var candidate: PsiElement = this

                while (candidate is RsElseBranch || candidate is RsIfExpr) {
                    candidate.leftSiblings.filterIsInstance<RsCondition>().forEach { set.add(it) }
                    candidate = candidate.parent
                }

                return set.any {
                    it.orPats?.patList?.all { pat -> pat?.isIrrefutable ?: false } ?: false
                }
            }
    }
```

Kobzol commented:
I created a new PR for `RsPatConst` here: https://github.com/intellij-rust/intellij-rust/pull/5199. I'll wait for that PR to be merged and also try to add some basic constant evaluation support.

Kobzol commented:
I rebased onto master with #5199 and added basic constant evaluation. I noticed that the evaluator is handling types in a peculiar way, if I have an expression `1 == 1`, it's of the `bool` type, so the evaluator is trying to search for `&&` or `||`, it doesn't find it so it fails. This could probably be improved in a different PR.

mchernyavsky commented:
bors retry

mchernyavsky commented:
bors r+
